### PR TITLE
feat: openapi config override

### LINKED ISSUE #961: feat: override openapi json
let the user override the openapi json file. refrain from generating, if:

a) the file is present somewhere, OR
b) the user passes it as a param to Robyn()

### PR DESCRIPTION
**Pull request by VishnuSanal**
## Description

This PR fixes #961

## Summary

This PR does add openapi config override feat

## PR Checklist

Please ensure that:

- [x] The PR contains a descriptive title
- [x] The PR contains a descriptive summary of the changes
- [x] You build and test your changes before submitting a PR.
- [x] You have added relevant documentation
- [ ] You have added relevant tests. We prefer integration tests wherever possible

## Pre-Commit Instructions:

- [x] Ensure that you have run the [pre-commit hooks](https://github.com/sparckles/robyn#%EF%B8%8F-to-develop-locally) on your PR.

### CODE REVIEW

**Comment by VishnuSanal** [comment]
how does this work with subrouters?

**Comment by sansyrox** [comment]
> how does this work with subrouters? The user would already have all endpoints in the openapi.json, I presume. it currently handles them.

@VishnuSanal , yes. Subrouters should not have the ability to override openapi

File: docs_src/src/pages/documentation/api_reference/openapi.mdx [hunk 1]
```diff
@@ -11,6 +11,10 @@ Out of the box, the following endpoints are setup for you:
 - `/docs` The Swagger UI
 - `/openapi.json` The JSON Specification
 
+To use a custom openapi configuration, you can:
+
+    - Place the `openapi.json` config file in the root directory.
+    - Or, pass the file path to the `openapi_json_path` parameter in the `Robyn()` constructor. (the parameter gets priority over the file).
```
Comment (sansyrox) [inline docs_src/src/pages/documentation/api_reference/openapi.mdx:17]: @VishnuSanal , can you rename this to `openapi_file_path`

File: docs_src/src/pages/documentation/example_app/openapi.mdx [hunk 2]
```diff
@@ -13,6 +13,11 @@ Out of the box, the following endpoints are setup for you:
 
 However, if you don't want to generate the OpenAPI docs, you can disable it by passing `--disable-openapi` flag while starting the application.
 
+To use a custom openapi configuration, you can:
+
+    - Place the `openapi.json` config file in the root directory.
+    - Or, pass the file path to the `openapi_json_path` parameter in the `Robyn()` constructor. (the parameter gets priority over the file).
```
Comment (sansyrox) [inline docs_src/src/pages/documentation/example_app/openapi.mdx:19]: same here. and in the entire pr

File: robyn/__init__.py [hunk 3]
```diff
@@ -583,7 +589,7 @@ def configure_authentication(self, authentication_handler: AuthenticationHandler
 
 class SubRouter(Robyn):
     def __init__(self, file_object: str, prefix: str = "", config: Config = Config(), openapi: OpenAPI = OpenAPI()) -> None:
-        super().__init__(file_object, config, openapi)
+        super().__init__(file_object=file_object, config=config, openapi=openapi)
```
Comment (sansyrox) [inline robyn/__init__.py:592]: Hey @VishnuSanal , now that I think of it. Can we remove openapi from the subrouters?

File: robyn/openapi.py [hunk 4]
```diff
@@ -138,11 +139,15 @@ class OpenAPI:
 
     info: OpenAPIInfo = field(default_factory=OpenAPIInfo)
     openapi_spec: dict = field(init=False)
+    is_pre_configured: bool = False
```
Comment (sansyrox) [inline robyn/openapi.py:142]: we should name it something else. `is_pre_configured` is a wrong term. should name it in the direction of override.

File: robyn/openapi.py [hunk 5]
```diff
@@ -163,6 +168,9 @@ def add_openapi_path_obj(self, route_type: str, endpoint: str, openapi_name: str
         @param handler: Callable the handler function for the endpoint
         """
 
+        if self.is_pre_configured:
+            return
+
```
Comment (sansyrox) [inline robyn/openapi.py:173]: same here.

File: robyn/__init__.py (hunk 3 above)
Comment (VishnuSanal) [reply robyn/__init__.py:592]: > Can we remove openapi from the subrouters?

noooo!! how does endpoints given via `@subrouter.METHOD` work? ref: https://robyn.tech/documentation/example_app/openapi#how-does-it-work-with-subrouters

File: robyn/__init__.py [hunk 6]
```diff
@@ -51,6 +52,11 @@ def __init__(
         self.dependencies = dependencies
         self.openapi = openapi
 
+        if openapi_file_path:
+            openapi.set_json_spec(self.directory_path + "/" + openapi_file_path)
+        elif os.path.isfile(self.directory_path + "/openapi.json"):
+            openapi.set_json_spec(self.directory_path + "/openapi.json")
+
```
Comment (sansyrox) [inline robyn/__init__.py:59]: @VishnuSanal , could you explain the logic here? What is this supposed to do? 

It should just be a file.

File: robyn/__init__.py (hunk 3 above)
Comment (sansyrox) [reply robyn/__init__.py:592]: Alright, makes sense.

File: robyn/openapi.py [hunk 7]
```diff
@@ -138,11 +139,15 @@ class OpenAPI:
 
     info: OpenAPIInfo = field(default_factory=OpenAPIInfo)
     openapi_spec: dict = field(init=False)
+    openapi_file_override: bool = False  # denotes whether running in the pre-configured file mode or not.
```
Comment (sansyrox) [inline robyn/openapi.py:142]:
```suggestion
    openapi_file_override: bool = False  # denotes whether there is an override or not.
```

File: robyn/__init__.py [hunk 8]
```diff
@@ -51,6 +52,11 @@ def __init__(
         self.dependencies = dependencies
         self.openapi = openapi
 
+        if openapi_file_path:
+            openapi.set_json_spec(self.directory_path + "/" + openapi_file_path)
+        elif os.path.isfile(self.directory_path + "/openapi.json"):
+            openapi.set_json_spec(self.directory_path + "/openapi.json")
```
Comment (sansyrox) [inline robyn/__init__.py:58]: and use pathlib to join the paths or os.path.join. `/` won't work on windows.

Comment (sansyrox) [reply robyn/__init__.py:58]: Can you please add integration tests for this?

**Review by sansyrox — COMMENTED** [review]
Have some suggestions @VishnuSanal

File: robyn/openapi.py [hunk 9]
```diff
@@ -393,6 +405,16 @@ def get_schema_object(self, parameter: str, param_type: Any) -> dict:
 
         return properties
 
+    def set_json_spec(self, openapi_json_spec_path: str):
```
[the remark this answers is not in the document]
Comment (sansyrox) [reply robyn/openapi.py:408]: @VishnuSanal , needs work here.

File: robyn/__init__.py (hunk 6 above)
Comment (VishnuSanal) [reply robyn/__init__.py:59]: this should invoke the override function if the path param is set or if the file i9s present in the pre-defnked path

**Review by sansyrox — APPROVED** [review]
LGTM! 👍

**Pull request merged**